### PR TITLE
User Applications selection using CMake

### PR DESCRIPTION
**Pull request by JF002**
Use CMake's configure_file() functionality to generate the list of User Applications.

All the apps included in current versions of InfiniTime are enabled by default, but this can now be overridden by setting variables ENABLE_APP_XXX to True or False.

CMake CMP0140 is set to NEW to enable the return PROPAGATE functionality.

### CODE REVIEW

**Review by FintasticMan — CHANGES_REQUESTED** [review]
This looks good! It allows a lot more freedom for building with whatever apps you want. I would say we should add this to the Apps.md documentation before merging, as you now need to register new apps in more places.

**Comment by JF002** [comment]
> This looks good! It allows a lot more freedom for building with whatever apps you want. I would say we should add this to the Apps.md documentation before merging, as you now need to register new apps in more places.

Thanks! I've just updated the doc to add instructions related to CMake.

**Review by FintasticMan — APPROVED** [review]

**Comment by FintasticMan** [comment]
It seems like InfiniSim will also need to include these same option declarations. @NeroBurner do you know if it would be possible to have InfiniSim get the options from InfiniTime?

File: CMakeLists.txt
Comment (NeroBurner) [inline CMakeLists.txt:37]: We can move those options and the Apps.h.in file into its own directory, create a CMake header only library target and link that to everywhere we need it. Because it is in its own folder I can easily include it in InfiniSim afterwards. Like we do for the fonts and images

File: CMakeLists.txt
Comment (NeroBurner) [inline CMakeLists.txt:89]: Not enabled, but comment in PR sais it is

Comment (FintasticMan) [reply CMakeLists.txt:89]: The version of CMake in the docker container is too old for this. See the second commit in this PR.

**Comment by minacode** [comment]
I don't know much about CMake, but could we put all optional apps in a separate directory and then autogenerate the CMake options?

```
.../displayapp
    apps/
        music/
            music.cpp
            music.h
        twos/
            twos.cpp
            twos.h
        ...
```

**Comment by FintasticMan** [comment]
I would imagine that that would be possible, and it would improve the dev experience quite a lot, not having to remember every single place to add a new app 😄

**Comment by FintasticMan** [comment]
One thing I've just thought of is that this doesn't allow choosing the order the apps appear in the app list. I think that that is quite an important feature, so we should try to figure out a way to get that to work.

One simple way would just be to have a string CMake parameter that gets placed in UserAppTypes.

File: CMakeLists.txt
Comment (JF002) [reply CMakeLists.txt:37]: I'll have a look at this!

**Comment by JF002** [comment]
@minacode 
> I don't know much about CMake, but could we put all optional apps in a separate directory and then autogenerate the CMake options?

It might be possible, yes. Actually, I would like InfiniTime to support User Applications coming from external repositories. This would allow maintainers of those user apps to maintain their apps the way they want, and to reduce the load on InfiniTime maintainers. I think we can do this in another future PR though :)

@FintasticMan
 > One thing I've just thought of is that this doesn't allow choosing the order the apps appear in the app list. I think that that is quite an important feature, so we should try to figure out a way to get that to work.

Good point! I'll try to find a solution for this!

**Review by FintasticMan — APPROVED** [review]

File: CMakeLists.txt
Comment (JF002) [reply CMakeLists.txt]: Done  in [c200816](https://github.com/InfiniTimeOrg/InfiniTime/pull/1928/commits/c200816416e09b890237625ab27d947904843eae) and integrated in InfiniSim here : https://github.com/InfiniTimeOrg/InfiniSim/pull/134

File: CMakeLists.txt
Comment (NeroBurner) [inline CMakeLists.txt:48]: I'd move this variable definition into src/displayapp/apps/CMakeLists.txt. then we'll have one place to update the default apps list and won't need to change the default list in InfiniSim as well.

Otherwise I'd need to update InfiniSim everytime the app list is changed to match InfiniTimes default

Comment (JF002) [reply CMakeLists.txt]: Good point, done!

**Comment by JF002** [comment]
@NeroBurner I'll apply the same change to the watchface selection when you confirm this is ok for you :)

**Comment by mark9064** [comment]
What about apps that introduce settings pages specific to them? We don't actually have any right now (arguably steps), but some are proposed e.g background heart rate settings

**Comment by minacode** [comment]
Handling such dependencies would be a next step. Apps also come with custom BLE interfaces (e.g. music).

**Comment by JF002** [comment]
@mark9064 That's a good question! And I agree with @minacode, we can figure that out in a next step. In this current state, the worse that can happen is that settings will show a setting that no userapp will actually use, but the whole firmware will still be functional.

File: doc/code/Apps.md
Comment (NeroBurner) [inline doc/code/Apps.md:62]: trailing space

```suggestion
This list of applications is defined as a list `Apps` enum values named `UserAppTypes` in `Apps.h`.
```

File: doc/code/Apps.md
Comment (NeroBurner) [inline doc/code/Apps.md:162]: use one line per sentence, makes the diffs in future easier, and "user application**s**"

```suggestion
Since the list of **user** applications is generated by CMake, you need to add the variable `ENABLE_USERAPPS` to the command line of CMake.
This variable must be set with a string composed of an ordered list of the **user** applications that must be built into the firmware.
The items of the list are fields from the enumeration `Apps`.
Ex : build the firmware with 3 user applications : Alarm, Timer and MyApp (the applications will be listed in this specific order in the application menu).
```

File: src/displayapp/apps/CMakeLists.txt
Comment (NeroBurner) [inline src/displayapp/apps/CMakeLists.txt:12]: add newline to end of file

```suggestion
configure_file(${CMAKE_CURRENT_SOURCE_DIR}/Apps.h.in ${CMAKE_CURRENT_BINARY_DIR}/Apps.h)

```

**Review by NeroBurner — APPROVED** [review]
small typos, afterwards good to go from my side :tada:

**Pull request merged**
